features/prog_test.go

```diff
@@ -92,3 +95,85 @@ func TestHaveProgTypeInvalid(t *testing.T) {
 		t.Fatalf("Expected os.ErrInvalid but was: %v", err)
 	}
 }
+
+func TestHaveProgHelper(t *testing.T) {
+	type testCase struct {
+		prog     ebpf.ProgramType
+		helper   asm.BuiltinFunc
+		expected error
+		version  string
+	}
+
+	// Referecing linux kernel commits to track the kernel version required to pass these test-cases
+	// These test-cases are derived from libbpf's selftests and helper/prog combinations that are
+	// probed for in cilium/cilium
+	// Still missing since those helpers are not available in the lib yet, are:
+	// - Kprobe, GetBranchSnapshot
+	// - SchedCLS, SkbSetTstamp
+
+	testCases := []testCase{
+		{ebpf.Kprobe, asm.FnMapLookupElem, nil, "3.19"},                     // @d0003ec
+		{ebpf.Kprobe, asm.FnKtimeGetCoarseNs, ebpf.ErrNotSupported, "5.16"}, // @5e0bc30
+		{ebpf.SocketFilter, asm.FnKtimeGetCoarseNs, nil, "5.11"},            // @d055126
+		{ebpf.SchedCLS, asm.FnSkbVlanPush, nil, "4.3"},                      // @4e10df9
+		{ebpf.Kprobe, asm.FnSysBpf, ebpf.ErrNotSupported, "5.14"},           // @79a7f8b
+		{ebpf.Syscall, asm.FnSysBpf, nil, "5.14"},                           // @79a7f8b
+		{ebpf.XDP, asm.FnJiffies64, nil, "5.5"},                             // @5576b99
+		{ebpf.XDP, asm.FnKtimeGetBootNs, nil, "5.7"},                        // @71d1921
+		{ebpf.SchedCLS, asm.FnSkbChangeHead, nil, "4.10"},                   // @3a0af8f
+		{ebpf.SchedCLS, asm.FnRedirectNeigh, nil, "5.10"},                   // @b4ab314
+		{ebpf.SchedCLS, asm.FnSkbEcnSetCe, nil, "5.1"},                      // @f7c917b
+		{ebpf.SchedACT, asm.FnSkAssign, nil, "5.6"},                         // @cf7fbe6
+		{ebpf.SchedACT, asm.FnFibLookup, nil, "4.18"},                       // @87f5fc7
+		{ebpf.CGroupSockAddr, asm.FnGetCgroupClassid, nil, "5.10"},          // @b426ce8
+		{ebpf.CGroupSockAddr, asm.FnGetsockopt, nil, "4.15"},                // @cd86d1f
+		{ebpf.CGroupSockAddr, asm.FnSkLookupTcp, nil, "4.20"},               // @6acc9b4
+		{ebpf.CGroupSockAddr, asm.FnGetNetnsCookie, nil, "5.7"},             // @f318903
+		{ebpf.CGroupSock, asm.FnGetNetnsCookie, nil, "5.7"},                 // @f318903
+
```

Comment:
Could we add one or two more expected `ErrNotSupported` combos from older (4.x) kernels? This would ensure our detection strategy is sound throughout the history. The current negative tests only run on 5.14 and above (modulo `TestHaveProgHelperUnsupported`, but that one might trip early user input checks in the syscall handler)

Reply:
Added:
* `{ebpf.Kprobe, asm.FnSkbVlanPush, ebpf.ErrNotSupported, "4.3"}, // 4e10df9a60d9`
* `{ebpf.Kprobe, asm.FnFibLookup, ebpf.ErrNotSupported, "4.18"},  // 87f5fc7e48dd`